### ISSUE TITLE
ImageSharp.Web always returning original image when a static content folder is specified

### ISSUE THREAD

**Issue by sophisma** [SixLabors/ImageSharp.Web#135]
### Description
Hi I'm trying ImageSharp.Web for the first time so, bear with me.
I created a fresh .Net Core 2.2 project and added the default configuration like you have on your documentation. In this scenario it works like it's supposed to but as soon as you change your static content folder it always returns the original unmodified image.

### Steps to Reproduce
Scenario 1 (this works).
Use this Configure method (ommited the ConfigureServices method).

        public void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }
            else
            {
                app.UseHsts();
            }

            app.UseHttpsRedirection();
            app.UseMvc();

            app.UseImageSharp();
            app.UseStaticFiles();
        }

Scenario 2 (this always returns the original image).

        public void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }
            else
            {
                app.UseHsts();
            }

            app.UseHttpsRedirection();
            app.UseMvc();

            app.UseImageSharp();
            app.UseStaticFiles(new StaticFileOptions
            {
                FileProvider = new PhysicalFileProvider("c:\\ImageSharpTest"),
                RequestPath = "/StaticFiles"
            });
        }

Checked the folder permissions and I have read and write permissions.

### System Configuration
- ImageSharp.Web version: 1.0.1
- Environment (Windows 10, VisualStudio 2019):
- .NET Framework version: .Net Core 2.2

**Comment by JimBobSquarePants** [comment]
The `PhysicalFileSystemProvider` is independent of the `StaticFileMiddleware` so changing the settings doesn't affect the provider in any way. 

However, it expects your source file to be in the `wwwroot` folder. Since you are now requesting files from a different location you need to implement your own `IImageProvider` that is aware of the new source file location in order to process the files. I would suggest using the existing provider source code as a basis to develop yours from. 

In short, the original is being returned because there are no matching providers for your new location so the request is delegated to the `StaticFileMiddleware`.

**Comment by sophisma** [comment]
Thank you for the quick reply and for putting me in the right direction.
I've managed to get the images from a different folder.
Just another quick question.
As you made perfectly clear in the previous post PhysicalFileSystemProvider and StaticFileMiddleware are different things.
Using StaticFileMiddleware I can define a Request Path for static files:

            app.UseStaticFiles(new StaticFileOptions
            {
                RequestPath = "/StaticFiles"
            });

Is there a similar thing I can do in ImageSharp.Web?
Currently I'm using both and I can access the images by accessing them like this:

https://localhost:44399/StaticFiles/image.png

If I put parameters in the URL I'm still getting the original one, however if I access them like this:

https://localhost:44399/image.png?height=73

It works (without /StaticFiles in the url).
I've been looking at the documentation but haven't found a way to make it work.
Could you point me to the right direction on how to add it to the URL?

**Comment by JimBobSquarePants** [comment]
I'm not sure you're fully understanding what I said before. 

Given two images:

- `<your_site>/StaticFiles/A.png?width=100`
- `<your_site>/B.png?width=100`

Regardless of custom changes to the static file middleware the `A.png` will be ignored by ImageSharp.Web and `B.png` will be processed.

This is because the `PhysicalFileSystemProvider` has no knowledge of the custom directory you are serving the image from but will serve images from the wwwroot folder based upon the given path.

https://github.com/SixLabors/ImageSharp.Web/blob/b72064b3b8cb8b883f8310c86b6d7e5643d80ad3/src/ImageSharp.Web/Providers/PhysicalFileSystemProvider.cs#L43-L45

Example paths that will work out of the box and their physical location.

| URL | Physical Location |
|----------|----------|
| `https://<your-site>/B.png` | `C:\your-site\wwwroot\B.png`|
| `https://<your-site>/subfolder/B.png` | `C:\your-site\wwwroot\subfolder\B.png`|

You can see from the .NET source that their `PhysicalFileProvider` simply pieces together a path based on the provider root.
https://source.dot.net/#Microsoft.Extensions.FileProviders.Physical/PhysicalFileProvider.cs,fcd614c88c2cc333,references

In order to capture files from a custom directory not rooted in the `wwwroot` folder you need to implement your own version of `IImageProvider` as I described before and then [register](https://docs.sixlabors.com/api/ImageSharp.Web/SixLabors.ImageSharp.Web.DependencyInjection.ImageSharpBuilderExtensions.html#SixLabors_ImageSharp_Web_DependencyInjection_ImageSharpBuilderExtensions_AddProvider__1_SixLabors_ImageSharp_Web_DependencyInjection_IImageSharpBuilder_) that provider on startup. This is the same requirement for any custom image sources.

Note: The provider collection should be [cleared](https://docs.sixlabors.com/api/ImageSharp.Web/SixLabors.ImageSharp.Web.DependencyInjection.ImageSharpBuilderExtensions.html#SixLabors_ImageSharp_Web_DependencyInjection_ImageSharpBuilderExtensions_ClearProviders_SixLabors_ImageSharp_Web_DependencyInjection_IImageSharpBuilder_) and the built in `PhysicalFileSystemProvider` should be registered after any custom providers should you still require it. 

I would copy the code from the `PhysicalFileSystemProvider` in this repository substituting the file provider set in the constructor with your own which has knowledge of the new path root.

**Comment by sophisma** [comment]
Let me explain what I did just to clear some confusion.
In your first reply you suggested me creating a new implementation of IImageProvider based on your PhysicalFileSystemProvider.
While reading the code I noticed that line of code you mentioned:

	this.fileProvider = environment.WebRootFileProvider;

This was actually very helpful because, since the file provider is being initiated with WebRootFileProvider, I could just change the default wwwroot of the webserver. This is perfectly acceptable in what I'm trying to achieve.
I did so by going to the Program.cs class and doing this:

	public static IWebHostBuilder CreateWebHostBuilder(string[] args) =>
            WebHost.CreateDefaultBuilder(args)
                .UseStartup<Startup>()
                .UseWebRoot("c:\\ImageSharpTest");

So now when I initiate both PhysicalFileSystemProvider and StaticFileMiddleware like this,

	app.UseImageSharp();
        app.UseStaticFiles();

they are both pointing at the same physical location ("c:\ImageSharpTest") and can be accessed like this:

https://<your-site>/A.png
https://<your-site>/A.png?width=30

This is because when you access https://<your-site>/ it resolves to your wwwroot folder.

Now, if I initialize them like this,

	app.UseImageSharp();
        app.UseStaticFiles(new StaticFileOptions
        {
            RequestPath = "/StaticFiles"
        });

they are both pointing at the same physical location ("c:\ImageSharpTest"), de only difference is in the routing on the controller.
There is no "StaticFiles" folder on the file system.
When I access

https://<your-site>/StaticFiles/A.png

this gets resolved to my wwwroot folder ("c:\ImageSharpTest").

But if I access

https://<your-site>/StaticFiles/A.png?width=10

This doesn't get resolved because PhysicalFileSystemProvider is not aware of that routing.
What I'm trying to figure out is if is there any way I can make this routing work on ImageSharp.Web.

**Comment by deanmarcussen** [comment]
> This doesn't get resolved because ImageSharp.Web is not aware of that routing.
> What I'm trying to figure out is if is there any way I can make this routing work on ImageSharp.Web.

As @JimBobSquarePants said above this won't work

With your current design and configuration it will resolve this url when serving a resized image
`https://<your-site>/A.png?width=10`

To serve under a different base / request path, i.e. `/StaticFiles/` you will need to implement an `IImageProvider` which is aware of your base / request path, of `/StaticFiles`.

Here is an example which does this, by default the base path used is `/media`

https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Media/Processing/MediaResizingFileProvider.cs

Refer to the `IsMatch` function https://github.com/OrchardCMS/OrchardCore/blob/cc677e2f103db525643fe999765c9ba10bd98313/src/OrchardCore.Modules/OrchardCore.Media/Processing/MediaResizingFileProvider.cs#L79

Note: You will still want to use the `PhysicalFileProvider` as a base, as the above file provider works with a different kind of file store.

**Comment by sophisma** [comment]
Thank you both for answering.
I managed to make it work with that example.

**Comment by sophisma** [comment]
I'm just putting this here just in case you want to support this out of the box.
I changed your PhysicalFileSystemProvider to work both ways, with and without request path.

namespace SixLabors.ImageSharp.Web.Providers
{
    /// <summary>
    /// Returns images stored in the local physical file system.
    /// </summary>
    public class CustomPhysicalFileSystemProvider : IImageProvider
    {
        /// <summary>
        /// The file provider abstraction.
        /// </summary>
        private readonly IFileProvider _fileProvider;

        /// <summary>
        /// Contains various format helper methods based on the current configuration.
        /// </summary>
        private readonly FormatUtilities _formatUtilities;

        /// <summary>
        /// Request base request path.
        /// </summary>
        private readonly PathString _requestPath;

        /// <summary>
        /// A match function used by the resolver to identify itself as the correct resolver to use.
        /// </summary>
        private Func<HttpContext, bool> _match;

        /// <summary>
        /// Initializes a new instance of the <see cref="PhysicalFileSystemProvider"/> class.
        /// </summary>
        /// <param name="environment">The environment used by this middleware.</param>
        /// <param name="formatUtilities">Contains various format helper methods based on the current configuration.</param>
        public CustomPhysicalFileSystemProvider(IHostingEnvironment environment, FormatUtilities formatUtilities, IOptions<CustomPhysicalFileSystemProviderOptions> mediaOptions)
        {
            Guard.NotNull(environment, nameof(environment));
            Guard.NotNull(environment.WebRootFileProvider, nameof(environment.WebRootFileProvider));
            this._fileProvider = environment.WebRootFileProvider;
            this._formatUtilities = formatUtilities;

            this._requestPath = mediaOptions.Value.RequestPath;
        }

        /// <inheritdoc/>
        public ProcessingBehavior ProcessingBehavior { get; } = ProcessingBehavior.CommandOnly;

        /// <inheritdoc/>
        public Func<HttpContext, bool> Match
        {
            get { return string.IsNullOrEmpty(_requestPath) ? _ => true : _match ?? IsMatch; }

            set => _match = value;
        }

        /// <inheritdoc/>
        public bool IsValidRequest(HttpContext context) => this._formatUtilities.GetExtensionFromUri(context.Request.GetDisplayUrl()) != null;

        /// <inheritdoc/>
        public Task<IImageResolver> GetAsync(HttpContext context)
        {
            // Remove assets request path if it's set.
            string path = string.IsNullOrEmpty(_requestPath) ? context.Request.Path.Value : context.Request.Path.Value.Substring(_requestPath.Value.Length);

            // Path has already been correctly parsed before here.
            IFileInfo fileInfo = this._fileProvider.GetFileInfo(path);

            // Check to see if the file exists.
            if (!fileInfo.Exists)
            {
                return Task.FromResult<IImageResolver>(null);
            }

            var metadata = new ImageMetadata(fileInfo.LastModified.UtcDateTime, fileInfo.Length);
            return Task.FromResult<IImageResolver>(new PhysicalFileSystemResolver(fileInfo, metadata));
        }

        private bool IsMatch(HttpContext context)
        {
            if (!context.Request.Path.StartsWithNormalizedSegments(_requestPath, StringComparison.OrdinalIgnoreCase))
            {
                return false;
            }

            return true;
        }
    }
}

I added this class.

namespace SixLabors.ImageSharp.Web.Providers
{
    public class CustomPhysicalFileSystemProviderOptions
    {
        public string RequestPath { get; set; }
    }
}

You can initialize it like this:

services.AddImageSharp()
    .RemoveProvider<PhysicalFileSystemProvider>()
    .AddProvider<CustomPhysicalFileSystemProvider>()
    .Configure<CustomPhysicalFileSystemProviderOptions>(options =>
                   options.RequestPath = "/RequestPath");
            
If you update your provider you only need the configure part obviously.

**Comment by creativesuspects** [comment]
@JimBobSquarePants I'm also using a static content folder outside of wwwroot on an Umbraco 9 project. I've already implemented a custom IImageProvider, but I'm a bit confused as to how (or where) I should tell ImageSharp to use the custom IImageProvider since ImageSharp is already being registered as middleware by Umbraco itself. Any ideas?

**Comment by JimBobSquarePants** [comment]
Ask Umbraco.

**Comment by zanyar3** [comment]
> I'm just putting this here just in case you want to support this out of the box. I changed your PhysicalFileSystemProvider to work both ways, with and without request path. I also added an option to set a custom content folder path.
> 
> ```
> namespace SixLabors.ImageSharp.Web.Providers
> {
>     /// <summary>
>     /// Returns images stored in the local physical file system.
>     /// </summary>
>     public class CustomPhysicalFileSystemProvider : IImageProvider
>     {
>         /// <summary>
>         /// The file provider abstraction.
>         /// </summary>
>         private readonly IFileProvider _fileProvider;
> 
>         /// <summary>
>         /// Contains various format helper methods based on the current configuration.
>         /// </summary>
>         private readonly FormatUtilities _formatUtilities;
> 
>         /// <summary>
>         /// Request base request path.
>         /// </summary>
>         private readonly PathString _requestPath;
> 
>         /// <summary>
>         /// A match function used by the resolver to identify itself as the correct resolver to use.
>         /// </summary>
>         private Func<HttpContext, bool> _match;
> 
>         /// <summary>
>         /// Initializes a new instance of the <see cref="PhysicalFileSystemProvider"/> class.
>         /// </summary>
>         /// <param name="environment">The environment used by this middleware.</param>
>         /// <param name="formatUtilities">Contains various format helper methods based on the current configuration.</param>
>         public CustomPhysicalFileSystemProvider(IHostingEnvironment environment, FormatUtilities formatUtilities, IOptions<CustomPhysicalFileSystemProviderOptions> mediaOptions)
>         {
>             Guard.NotNull(environment, nameof(environment));
>             Guard.NotNull(environment.WebRootFileProvider, nameof(environment.WebRootFileProvider));
>             this._fileProvider = string.IsNullOrEmpty(mediaOptions.Value.ContentPath) ? environment.WebRootFileProvider : new PhysicalFileProvider(mediaOptions.Value.ContentPath);
>             this._formatUtilities = formatUtilities;
> 
>             this._requestPath = mediaOptions.Value.RequestPath;
>         }
> 
>         /// <inheritdoc/>
>         public ProcessingBehavior ProcessingBehavior { get; } = ProcessingBehavior.CommandOnly;
> 
>         /// <inheritdoc/>
>         public Func<HttpContext, bool> Match
>         {
>             get { return string.IsNullOrEmpty(_requestPath) ? _ => true : _match ?? IsMatch; }
> 
>             set => _match = value;
>         }
> 
>         /// <inheritdoc/>
>         public bool IsValidRequest(HttpContext context) => this._formatUtilities.GetExtensionFromUri(context.Request.GetDisplayUrl()) != null;
> 
>         /// <inheritdoc/>
>         public Task<IImageResolver> GetAsync(HttpContext context)
>         {
>             // Remove assets request path if it's set.
>             string path = string.IsNullOrEmpty(_requestPath) ? context.Request.Path.Value : context.Request.Path.Value.Substring(_requestPath.Value.Length);
> 
>             // Path has already been correctly parsed before here.
>             IFileInfo fileInfo = this._fileProvider.GetFileInfo(path);
> 
>             // Check to see if the file exists.
>             if (!fileInfo.Exists)
>             {
>                 return Task.FromResult<IImageResolver>(null);
>             }
> 
>             var metadata = new ImageMetadata(fileInfo.LastModified.UtcDateTime, fileInfo.Length);
>             return Task.FromResult<IImageResolver>(new PhysicalFileSystemResolver(fileInfo, metadata));
>         }
> 
>         private bool IsMatch(HttpContext context)
>         {
>             if (!context.Request.Path.StartsWithNormalizedSegments(_requestPath, StringComparison.OrdinalIgnoreCase))
>             {
>                 return false;
>             }
> 
>             return true;
>         }
>     }
> }
> ```
> 
> I added this class.
> 
> ```
> namespace SixLabors.ImageSharp.Web.Providers
> {
>     public class CustomPhysicalFileSystemProviderOptions
>     {
>         public string RequestPath { get; set; }
>         public string ContentPath { get; set; }
>     }
> }
> ```
> 
> You can initialize it like this:
> 
> ```
> services.AddImageSharp()
>     .RemoveProvider<PhysicalFileSystemProvider>()
>     .AddProvider<CustomPhysicalFileSystemProvider>()
>     .Configure<CustomPhysicalFileSystemProviderOptions>(options =>
>     {
>         options.RequestPath = "/YourRequestPath";
>         options.ContentPath = "c:\\YourContentPath\\";
>     });
> ```
> 
> If you update your provider you only need the configure part obviously, not the Remove/Add providers part. One thing you have to be aware though is that I took the StartsWithNormalizedSegments method from here: https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore/OrchardCore.Abstractions/Routing/PathStringExtensions.cs

@sophisma  Thanks it is work successfully, but not work for cache folder 

```
services.AddImageSharp()
               .RemoveProvider<PhysicalFileSystemProvider>()
               .AddProvider<CustomPhysicalFileSystemProvider>()
               .Configure<CustomPhysicalFileSystemProviderOptions>(options =>
               {
                   options.ContentPath = EnvironmentConfig.FilePublicPath;
                   options.RequestPath = "/public";
               })
               .Configure<PhysicalFileSystemCacheOptions>(options =>
               {
                   options.CacheFolder = "file-caches";
               });
```
### Don't create `file-caches` folder

**Comment by sophisma** [comment]
I didn't mess with the cache folder.
Check if the application has permissions to create it or maybe you'll have to create it by hand?

**Comment by sophisma** [comment]
I just remembered that in my case it creates the cache folder inside my specified content folder automatically.
It really might be a question of the server not having write permissions for the content folder.

**Comment by JimBobSquarePants** [comment]
Please don’t comment on closed issues.
